Makefile

```diff
@@ -349,3 +349,27 @@ integration-test:
 	else\
 		$(GINKGO) -v $(TEST_ARGS) --fail-fast --focus-file=$(TEST_FILE) test/integration;\
 	fi
+
+#hosted
+
+.PHONY: bootstrap-hosted
+bootstrap-hosted: clone-hosted install-hosted copy-config install-crds install-resources kind-deploy-policy-framework kind-deploy-policy-controllers
```

Comment:
`kind-deploy-policy-framework` and `kind-deploy-policy-controllers` deploy the controllers in normal mode. Refer back to our meeting recording to see the difference. In particular, the `ManagedClusterAddOn` object needs to have the `addon.open-cluster-management.io/hosting-cluster-name` annotation set. Additionally, it needs `installNamespace` to be set to something else.

Reply:
Additionally, it seems that these tests are bypassing the governance-policy-addon-controller altogether, so you'll need to deploy the controllers first by installing the `governance-policy-addon-controller` on the hosted mode hub.

Also, remember that you need to create the secret with the Kubeconfig like I showed you in the recording for the controller on the management cluster to connect to the hosted/managed cluster.

Reply:
If we use ManagedClusterAddOn like you showed before, we cannot pass the e2e test, because, for example, e2e test check policy copy in managed cluster. ManagedClusterAddOn way only creates something under policy and configurationpolicy

Reply:
I suggest, if we want to test hosted mode in the way of "managedclusteraddon" as you showed me before, we need to write separate e2e tests for hosted mode

Reply:
> these tests are bypassing the governance-policy-addon-controller altogether, so you'll need to deploy the controllers first by installing the governance-policy-addon-controller on the hosted mode hub.

I wonder if that should be a separate PR/effort, possibly before we try to run these tests with a hosted cluster. 

> e2e test check policy copy in managed cluster

Right - currently, there are tests (and common test functions) that check that the replicated/propagated/child policies are copied to a cluster. For example: [here](https://github.com/stolostron/governance-policy-framework/blob/d6432a6757ea9f99bf750416778f359909658174/test/common/policy_utils.go#L107), [here](https://github.com/stolostron/governance-policy-framework/blob/main/test/common/policy_utils.go#L149), and [here](https://github.com/stolostron/governance-policy-framework/blob/main/test/configuration_policy_prune.go#L145). There are also tests that check for managed objects on a cluster, for example: [here](https://github.com/stolostron/governance-policy-framework/blob/main/test/configuration_policy_prune.go#L163), and [here](https://github.com/stolostron/governance-policy-framework/blob/main/test/e2e/configuration_policy_test.go#L117). Currently in both of those situations, they check the managed cluster. But, in hosted mode, the policies will be created on the "hosting cluster" (maybe usually the hub), while the managed objects are created on the "hosted cluster". My thought for this is that we have 3 kubeconfigs/clients, where currently we have 2:
1. A hub kubeconfig, like we currently have. Used for root policies, and verifying overall status.
2. A "hosting cluster" kubeconfig, which for our 'traditional' test runs would be the same as the managed kubeconfig. For new 'hosted mode' test runs, based on what I've heard here, this would actually be the hub cluster, but in the future we could potentially expand it to a separate cluster. This would be used for replicated policies and policy templates, and events.
3. A "target cluster" kubeconfig, which for our 'traditional' tests would be the same as the managed kubeconfig. This would be used to check that objects in a configuration policy were actually created, and for creating things like users and certificates to verify the functionality of iam/cert policies.

The test runs could always take all three of those via [configuration flags](https://github.com/stolostron/governance-policy-framework/blob/d6432a6757ea9f99bf750416778f359909658174/test/common/common.go#L48), and we could give the test runs the right values based on the type of the test. Then we would just need to update the current tests (see examples above) to use the correct clients where currently they always point to the managed cluster.

Reply:
@JustinKuli thanks for the detailed suggestion!

@yiraeChristineKim could you please pursue this approach?

Reply:
understand clear